File: UI/travle_core/lib/src/models/tour_insert_request.dart
import 'package:json_annotation/json_annotation.dart';

part 'tour_insert_request.g.dart';

/// An organizer's new tour (mirrors the backend `TourInsertRequest`). The
/// organizer comes from the JWT and is never sent. [destinationIds] is the ordered
/// itinerary — the list order becomes each stop's sort order — and every id must
/// reference an approved destination (verified server-side).
@JsonSerializable()
class TourInsertRequest {
  TourInsertRequest({
    required this.name,
    required this.description,
    required this.durationMinutes,
    required this.pricePerPerson,
    required this.capacity,
    required this.tourTypeId,
    this.bookingCutoffMinutes,
    this.destinationIds = const [],
  });

  final String name;
  final String description;
  final int durationMinutes;
  final double pricePerPerson;
  final int capacity;
  final int tourTypeId;

  /// Minutes before a departure that bookings and payments close for this tour.
  /// Null uses the platform default; 0 keeps a departure bookable until it starts.
  final int? bookingCutoffMinutes;

  final List<int> destinationIds;

  factory TourInsertRequest.fromJson(Map<String, dynamic> json) =>
      _$TourInsertRequestFromJson(json);

  Map<String, dynamic> toJson() => _$TourInsertRequestToJson(this);
}

File: Backend/Travle.Model/Requests/TourInsertRequest.cs
namespace Travle.Model.Requests
{
    /// <summary>
    /// An organizer's new tour. The organizer is taken from the JWT (never the client). Tour type is an
    /// FK id resolved server-side. <see cref="DestinationIds"/> is the ordered itinerary — the list
    /// order becomes each stop's <c>SortOrder</c> — and every id must reference an approved destination
    /// (verified in the service, since it needs the database).
    /// </summary>
    public class TourInsertRequest
    {
        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;

        public int DurationMinutes { get; set; }
        public decimal PricePerPerson { get; set; }

        /// <summary>Default group size seeded into new schedules.</summary>
        public int Capacity { get; set; }

        public int TourTypeId { get; set; }

        /// <summary>
        /// Minutes before a departure that bookings and payments close for this tour — the organizer's lead
        /// time to confirm or reject. Null uses the platform default; 0 keeps a departure bookable until it
        /// starts.
        /// </summary>
        public int? BookingCutoffMinutes { get; set; }

        /// <summary>Approved destinations to visit, in itinerary order (order = SortOrder). At least one.</summary>
        public List<int> DestinationIds { get; set; } = new List<int>();
    }
}
